apps/meteor/app/livechat/server/api/v1/contact.ts

```diff
@@ -41,29 +42,48 @@ API.v1.addRoute(
 
 API.v1.addRoute(
 	'omnichannel/contact.search',
-	{ authRequired: true },
+	{ authRequired: true, permissionsRequired: ['view-l-room'] },
 	{
 		async get() {
 			check(this.queryParams, {
 				email: Match.Maybe(String),
 				phone: Match.Maybe(String),
+				custom: Match.Maybe(String),
 			});
+			const { email, phone, custom } = this.queryParams;
 
-			const { email, phone } = this.queryParams;
-
-			if (!email && !phone) {
+			if (!email && !phone && !custom) {
 				throw new Meteor.Error('error-invalid-params');
 			}
+			let foundCF = {};
+			if (custom) {
+				const customObj = Object.fromEntries(Array.from(new URLSearchParams(custom)));
+
+				foundCF = Object.fromEntries(
+					(
+						await LivechatCustomField.find(
+							{ scope: 'visitor', searchable: true, _id: { $in: Object.keys(customObj) } },
+							{
+								projection: {
+									_id: 1,
+								},
+							},
+						).toArray()
+					).map(({ _id }) => [`livechatData.${_id}`, new RegExp(escapeRegExp(customObj[_id]), 'i')]),
+				);
+			}
 
 			const query = Object.assign(
 				{},
 				{
 					...(email && { visitorEmails: { address: email } }),
 					...(phone && { phone: { phoneNumber: phone } }),
+					...(Object.keys(foundCF).length && foundCF),
 				},
 			);
 
-			const contact = await LivechatVisitors.findOne(query);
+			// If no valid query items are provided, return null (to avoid returning a random visitor)
+			const contact = Object.keys(query).length ? await LivechatVisitors.findOne(query) : null;
```

Comment:
this is actually not required as it is already being checked on line 55
```suggestion
			const contact = await LivechatVisitors.findOne(query);
```